### DISCUSSION TITLE
Trying to draw MPV under egui

### DISCUSSION THREAD

**Discussion by StratusFearMe21** [emilk/egui#1641] [Q&A] [answered]
## The Situation

I'm trying to make my own MPV player using the power of `egui_glium`, however, it seems that egui adds a red/blue tint to the video, and doesn't actually draw anything.
![screenshot](https://user-images.githubusercontent.com/57533634/169087343-c6e4820e-05b1-42f7-85f8-47a4dd4ee5a1.png)
I don't know what I'm doing wrong, am I missing something? This is the redraw function in my code
```rust
let mut quit = false;
let needs_repaint = egui_glium.run(&display, |egui_ctx| {
    egui::SidePanel::left("my_side_panel").show(egui_ctx, |ui| {
        ui.heading("Hello World!");
        if ui.button("Quit").clicked() {
            quit = true;
        }
    });
});

*control_flow = if quit {
    glium::glutin::event_loop::ControlFlow::Exit
} else if needs_repaint {
    display.gl_window().window().request_redraw();
    glium::glutin::event_loop::ControlFlow::Poll
} else {
    glium::glutin::event_loop::ControlFlow::Wait
};

{
    use glium::Surface as _;
    let mut target = display.draw();

    // draw things behind egui here

    target.clear_color(0.0, 0.0, 0.0, 0.0);
    if let Some(render_context) = &render_context {
        render_context
            .render::<Display>(0, size.0 as _, size.1 as _, true)
            .expect("Failed to draw on glutin window");
    }

    egui_glium.paint(&display, &mut target);

    // draw things on top of egui here

    target.finish().unwrap();
}
```
I could be wrong but it may have to do with the fact that mpv expects "OpenGL standard defaults" (I'm not sure though because removing the `egui_glium.paint(&display, &mut target);` line allows the video to draw correctly, but no GUI), this is what it says in the docs
```c
/**
 * OpenGL backend
 * --------------
 *
 * This header contains definitions for using OpenGL with the render.h API.
 *
 * OpenGL interop
 * --------------
 *
 * The OpenGL backend has some special rules, because OpenGL itself uses
 * implicit per-thread contexts, which causes additional API problems.
 *
 * This assumes the OpenGL context lives on a certain thread controlled by the
 * API user. All mpv_render_* APIs have to be assumed to implicitly use the
 * OpenGL context if you pass a mpv_render_context using the OpenGL backend,
 * unless specified otherwise.
 *
 * The OpenGL context is indirectly accessed through the OpenGL function
 * pointers returned by the get_proc_address callback in mpv_opengl_init_params.
 * Generally, mpv will not load the system OpenGL library when using this API.
 *
 * OpenGL state
 * ------------
 *
 * OpenGL has a large amount of implicit state. All the mpv functions mentioned
 * above expect that the OpenGL state is reasonably set to OpenGL standard
 * defaults. Likewise, mpv will attempt to leave the OpenGL context with
 * standard defaults. The following state is excluded from this:
 *
 *      - the glViewport state
 *      - the glScissor state (but GL_SCISSOR_TEST is in its default value)
 *      - glBlendFuncSeparate() state (but GL_BLEND is in its default value)
 *      - glClearColor() state
 *      - mpv may overwrite the callback set with glDebugMessageCallback()
 *      - mpv always disables GL_DITHER at init
 *
 * Messing with the state could be avoided by creating shared OpenGL contexts,
 * but this is avoided for the sake of compatibility and interoperability.
 *
 * On OpenGL 2.1, mpv will strictly call functions like glGenTextures() to
 * create OpenGL objects. You will have to do the same. This ensures that
 * objects created by mpv and the API users don't clash. Also, legacy state
 * must be either in its defaults, or not interfere with core state.
 */
```
I'm not very knowledgeable in OpenGL, so that's probably a contributing factor

## Edit:
I've tried to do the same thing with `egui_glow`, and the effect is much less pronounced now
![screenshot](https://user-images.githubusercontent.com/57533634/169155935-a88cb62c-6570-434d-baf9-198f32c0c059.png)
However, I've noticed that the `egui` stream is much brighter than the `mpv` stream (caused by sRGB maybe?)

## Notes:
- If I draw `mpv` *after* `egui` the problem persists and no gui is drawn
- If I initialize `egui_glow`/`egui_glium`, as long as I never call `egui_glow.paint()`, the problem goes away but no gui
- Modifying the shaders in `egui_glow` to simplify the srgb => linear and vice versa conversions to simply
```glsl
vec4 srgba_from_linear(vec4 rgba) {
	return rgba * 255.0;
}

vec4 linear_from_srgba(vec4 srgba) {
	return srgba / 255.0;
}
```
results in a very similar brightness problem except from the UI itself.
![screenshot](https://user-images.githubusercontent.com/57533634/169163465-7aa71030-d31b-446e-bfd3-fc8a32d2f8b8.png)

**Comment by cdknight** [comment, marked as answer]
I found disabling `GL_FRAMEBUFFER_SRGB` (`gl.disable(glow::FRAMEBUFFER_SRGB)`) when trying my own experiment with rendering MPV to fix this issue, but I am not sure if the resulting egui will render the "correct" colours (highlighting my lack of knowledge about colour spaces). You can see my example code: https://github.com/cdknight/mpv-egui/
